### PR TITLE
Strengthen typing of Runtimes+Langauges and their support timelines.

### PR DESCRIPTION
**Pull request by inlined**
- Factors out Runtimes into a new file so that they can be imported into firebaseConfig.ts and have a single source of truth and prevent the error encountered in https://github.com/firebase/firebase-tools/issues/6774#issuecomment-1953399977
- Adds a formal concept of languages and the ability to discriminate runtimes per language
- Makes all information about a runtime table driven so there can never be a runtime without all necessary metadata
- Adds the ability to dynamically fetch the latest runtime for a language
- Adds support timelines for all runtimes
- Unifies runtime support checks across all runtimes/languages. There are now separate warnings for when deprecation is upcoming (90d), when a runtime is deprecated, and when a runtime is decommissioned.

### CODE REVIEW

File: src/deploy/functions/runtimes/python/index.ts [hunk 1]
```diff
@@ -30,9 +29,15 @@ export async function tryCreateDelegate(
     logger.debug("Customer code is not Python code.");
     return;
   }
-  const runtime = context.runtime ? context.runtime : LATEST_VERSION;
-  if (!runtimes.isValidRuntime(runtime)) {
-    throw new FirebaseError(`Runtime ${runtime} is not a valid Python runtime`);
+  const runtime = context.runtime ? context.runtime : supported.latest("python");
```
Comment (joehan) [inline src/deploy/functions/runtimes/python/index.ts:32]: Nit: `context.runtime ?? supported.latest("python")` is probably cleaner than a ternary

File: src/test/deploy/functions/checkIam.spec.ts [hunk 2]
```diff
@@ -21,7 +21,7 @@ const BINDING = {
 const SPEC = {
   region: "us-west1",
   project: projectNumber,
-  runtime: "nodejs14",
+  runtime: "nodejs14" as const,
```
Comment (joehan) [inline src/test/deploy/functions/checkIam.spec.ts:24]: TIL that `as const` satisfies enums without having to import them - good to know!

**Review by joehan — APPROVED** [review]
MInor nits, but lgtm!

File: src/test/deploy/functions/checkIam.spec.ts (hunk 2 above)
Comment (inlined) [reply src/test/deploy/functions/checkIam.spec.ts:24]: yes, "as const" will make a string literal type rather than just a string type.

**Pull request merged**